File: analysers/go/exp/metrics/cloc.go
package metrics

import (
	"encoding/json"
	"os/exec"
)

type FileClocStat struct {
	Header struct {
		ClocURL        string  `json:"cloc_url"`
		ClocVersion    string  `json:"cloc_version"`
		ElapsedSeconds float64 `json:"elapsed_seconds"`
		NFiles         int     `json:"n_files"`
		NLines         int     `json:"n_lines"`
		FilesPerSecond float64 `json:"files_per_second"`
		LinesPerSecond float64 `json:"lines_per_second"`
	} `json:"header"`
	Go struct {
		NFiles  int `json:"nFiles"`
		Blank   int `json:"blank"`
		Comment int `json:"comment"`
		Code    int `json:"code"`
	} `json:"Go"`
	Sum struct {
		Blank   int `json:"blank"`
		Comment int `json:"comment"`
		Code    int `json:"code"`
		NFiles  int `json:"nFiles"`
	} `json:"SUM"`
}

func fileCLOC(filename string) (fileCloc FileClocStat, err error) {
	fileCloc = FileClocStat{}
	var cmd = exec.Command("cloc", filename, "--json")
	output, err := cmd.Output()
	if err != nil {
		return
	}
	err = json.Unmarshal(output, &fileCloc)
	return
}

File: analysers/py/metrics/cloc.py
from dataclasses import dataclass
from dataclasses_json import dataclass_json
from subprocess import run

from logging import getLogger

log = getLogger(__name__)


@dataclass_json
@dataclass(frozen=True)
class Header:
    cloc_url: str
    cloc_version: str
    elapsed_seconds: float
    n_files: int
    n_lines: int
    files_per_second: float
    lines_per_second: float


@dataclass_json
@dataclass(frozen=True)
class Python:
    nFiles: int
    blank: int
    comment: int
    code: int


@dataclass_json
@dataclass(frozen=True)
class Sum:
    blank: int
    comment: int
    code: int
    nFiles: int


@dataclass_json
@dataclass(frozen=True)
class FileClocStat:
    header: Header
    Python: Python
    SUM: Sum


def _file_cloc(file_name: str) -> FileClocStat:
    cp = run(["cloc", file_name, "--json"], capture_output=True)
    # TODO: cloc is weird: 'Unable to read:  /home/zsoltk/git/linux-rag-t2/backend/src/adapters/ollama/client.pyk' but returns with '0'
    cp.check_returncode()
    if cp.returncode != 0:
        raise ChildProcessError(cp.stderr)
    return FileClocStat.from_json(cp.stdout)  # type: ignore
